### DISCUSSION TITLE
checking whether package can be installed... NOTE  Installation took CPU time `x` times elapsed time

### DISCUSSION THREAD

**Discussion by aitap** [r-devel/cran-cookbook#61] [Community Contributions]
**Problem**: Your package got a NOTE saying that during installation, it took more CPU time than elapsed time.

**Cause**: CRAN checks run on big, multi-core computers, running many `R CMD check` processes in parallel. Every process is allowed to use two CPU cores to exercise parallel code in examples and unit tests ([policy](https://cran.r-project.org/web/packages/policies.html): "If running a package uses multiple threads/cores it must never use more than two simultaneously"). Something that ran as part of the installation process (most likely called from `src/Makevars*`) started more than two child processes (or threads) at the same time.

* Example cause seen with Rust: running `cargo` without explicit `-j 2` uses all available CPUs by default [[1](https://stat.ethz.ch/pipermail/r-package-devel/2024q3/010973.html), [2](https://stat.ethz.ch/pipermail/r-package-devel/2024q4/011132.html)]
* Possible cause when using [CMake](https://cmake.org) and/or [Ninja](https://ninja-build.org/): when using `cmake -GNinja` and then `cmake --build`, `ninja` will by default use all available CPUs.
* Possible cause when using Make: some third-party software may call `` make -j`nproc` `` in its build scripts to speed up the build process.

**Solution**: Specific to the build system your package is using, but many modern build systems use the `-j` flag that had originated in Make:

 * For `cargo`, make sure to [specify `-j 2` any time your package is built on CRAN](https://github.com/extendr/rextendr/blob/c698bf3445a6160fd99fff574ef61fd442a39e36/inst/templates/cran/Makevars#L10).
 * For `cmake --build`, also make sure to specify `-j 2` at least on CRAN builds to avoid the underlying build system implicitly using all CPUs.
 * For Make, locate the explicit uses of the `-j` flag and make sure it doesn't exceed two at least on CRAN builds.

**Comment by Baltmann01** [comment]
Thanks for the contribution. This issue (to some extend) is already on our ToDo list. I suspect it will be added next week. 
I will look at your suggestions and include them in the recipe when writing it.

**Reply by aitap** [reply]
I think [I see it now](https://github.com/orgs/r-devel/projects/3?pane=issue&itemId=81474421). Will check the ToDo list next time I decide to file something here, thanks!

**Reply by Baltmann01** [reply]
Just as a heads-up: I just created a PR (#63)  about this issue. Thanks for all your suggestions! I added more details than I had planned due to this post :)
